File: Sorting Algorithms/shell_sort.cpp
#include<bits/stdc++.h>
using namespace std;

void display(int *array,int size){
    for(int i=0;i<size;i++){
        cout<<array[i]<<" ";
    }
    cout<<"\n";
}

void swapping(int &a, int &b){
    int temp;
    temp=a;
    a=b;
    b=temp;
}

void shellSort(int *arr,int n){
    int gap,j,k;
    for(gap=n/2;gap>0;gap=gap/2){
        for(j=gap;j<n;j++){
            for(k=j-gap;k>=0;k-=gap){
                if(arr[k+gap]>=arr[k])
                   break;
                else 
                  swapping(arr[k+gap],arr[k]);
            }
        }

    }
    
}

int main(){
    int n;
    cout<<"Enter the number of elements: ";
    cin>>n;
    int arr[n];
    cout<<"Enter elements: "<<"\n";
    for(int i=0;i<n;i++){
        cin>>arr[i];
    }
    cout<<"Array before Sorting: ";
    display(arr,n);
    shellSort(arr,n);
    cout<<"Array after Sorting: ";
    display(arr,n);

  
}

File: Sorting Algorithms/shell_sort.py
#function to implement the shell sort
def shellsort(arr,n):

    #calculate gap which is half of the length
    gap=n//2

    # Do a gapped insertion sort for this gap size.
    # The first gap elements a[0..gap-1] are already in gapped 
    # order keep adding one more element until the entire array
    # is gap sorted
    while(gap>0):

        #Loop iterating from gap to n-1
        for i in range(gap,n):
            temp=arr[i]  #value in arr[i] is assigned to temporary variable
            j=i   

            while j>=gap and arr[j-gap]>temp:
                arr[j]=arr[j-gap]  #value of arr[j-gap] is assigned to arr[j]
                j=j-gap   
            arr[j]=temp    #After loop terminates temporary value is assigned to arr[j]
        gap=gap//2      



n=int(input("Enter the no of elements:"))
arr=list(map(int,input("Enter the elements:").split()))
print("Before Shell Sort")
print(arr)
shellsort(arr,n)
print("After Shell Sort")
print(arr)